13_BitonicSort/main.cpp

```diff
@@ -0,0 +1,319 @@
+
+			IQueue* const queue = getComputeQueue();
+			uint32_t queueFamilyIndex = queue->getFamilyIndex();
+
+			deviceLocalBufferParams.queueFamilyIndexCount = 1;
+			deviceLocalBufferParams.queueFamilyIndices = &queueFamilyIndex;
+			deviceLocalBufferParams.size = sizeof(uint32_t) * elementCount * 2;  // *2 because we store (key, value) pairs
+			deviceLocalBufferParams.usage = nbl::asset::IBuffer::E_USAGE_FLAGS::EUF_TRANSFER_SRC_BIT | nbl::asset::IBuffer::E_USAGE_FLAGS::EUF_TRANSFER_DST_BIT | nbl::asset::IBuffer::E_USAGE_FLAGS::EUF_SHADER_DEVICE_ADDRESS_BIT;
+
+			m_deviceLocalBuffer = m_device->createBuffer(std::move(deviceLocalBufferParams));
+			auto mreqs = m_deviceLocalBuffer->getMemoryReqs();
+			mreqs.memoryTypeBits &= m_device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+			auto gpubufMem = m_device->allocate(mreqs, m_deviceLocalBuffer.get(), IDeviceMemoryAllocation::E_MEMORY_ALLOCATE_FLAGS::EMAF_DEVICE_ADDRESS_BIT);
+
+			m_deviceLocalBufferAddress = m_deviceLocalBuffer.get()->getDeviceAddress();
+		}
+
+		const nbl::asset::SPushConstantRange pcRange = {
+			.stageFlags = IShader::E_SHADER_STAGE::ESS_COMPUTE,
+			.offset = 0,
+			.size = sizeof(PushConstantData)
+		};
+
+		{
+			m_layout = m_device->createPipelineLayout({ &pcRange,1 });
+			IGPUComputePipeline::SCreationParams params = {};
+			params.layout = m_layout.get();
+			params.shader.shader = bitonicSortShader.get();
+			params.shader.entryPoint = "main";
+			params.shader.requiredSubgroupSize = static_cast<IPipelineBase::SUBGROUP_SIZE>(hlsl::findMSB(m_physicalDevice->getLimits().maxSubgroupSize));
+			params.cached.requireFullSubgroups = true;
+			if (!m_device->createComputePipelines(nullptr, { &params,1 }, &m_pipeline))
+				return logFail("Failed to create compute pipeline!\n");
+		}
+
+		const auto& deviceLimits = m_device->getPhysicalDevice()->getLimits();
+		m_alignment = core::max(deviceLimits.nonCoherentAtomSize, alignof(uint32_t));
+
+		m_timeline = m_device->createSemaphore(semaphoreValue);
+
+		IQueue* const queue = getComputeQueue();
+
+		const uint32_t inputSize = sizeof(uint32_t) * elementCount * 2;  // *2 because we store (key, value) pairs
+
+		const uint32_t AllocationCount = 1;
+
+		auto inputOffset = m_upStreamingBuffer->invalid_value;
+
+		std::chrono::steady_clock::time_point waitTill(std::chrono::years(45));
+		m_upStreamingBuffer->multi_allocate(waitTill, AllocationCount, &inputOffset, &inputSize, &m_alignment);
+
+		{
+			auto* const inputPtr = reinterpret_cast<uint32_t*>(reinterpret_cast<uint8_t*>(m_upStreamingBuffer->getBufferPointer()) + inputOffset);
+
+			// Generate random input data
+			unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+			std::mt19937 g(seed);
+
+			std::cout << "Input array: ";
+			for (uint32_t i = 0; i < elementCount; i++) {
+				uint32_t key = g() % 10000;
+				uint32_t value = i; // Use index as value for stable sorting
+				inputPtr[i * 2] = key;
+				inputPtr[i * 2 + 1] = value;
```

Comment:
again another reason to have a named struct so you're not doing this unreadable type punning